### PR TITLE
Tree behavior improvements

### PR DESCRIPTION
**Pull request by lorenzo**
This is a complete version of #8022

Also, this adds support for callables in `Query::set()`

### CODE REVIEW

File: src/ORM/Behavior/TreeBehavior.php [hunk 1]
```diff
@@ -76,6 +78,15 @@ class TreeBehavior extends Behavior
     ];
 
     /**
+     * {@inheritDoc}
+     */
+    public function initialize(array $config)
+    {
+        $this->_config['leftField'] = new IdentifierExpression($this->_config['left']);
+        $this->_config['rightField'] = new IdentifierExpression($this->_config['right']);
```
Comment (markstory) [inline src/ORM/Behavior/TreeBehavior.php]: What about runtime configuration changes?

File: src/ORM/Behavior/TreeBehavior.php [hunk 2]
```diff
@@ -843,17 +873,28 @@ protected function _getMax()
...
+        foreach ([$config['leftField'], $config['rightField']] as $field) {
             $query = $this->_scope($this->_table->query());
+            $exp = $query->newExpr();
+
+            $movement = clone $exp;
+            $movement ->add($field)->add("$shift")->type($dir);
+
+            $inverse = clone $exp;
+            $movement = $mark ?
+                $inverse->add($movement)->add('-1')->type('*') :
```
Comment (markstory) [inline src/ORM/Behavior/TreeBehavior.php]: Would the code be easier to read if it was `->type('*')->add('-1')`?

Comment (lorenzo) [inline src/ORM/Behavior/TreeBehavior.php]: good idea

File: src/ORM/Behavior/TreeBehavior.php (hunk 1 above)
Comment (lorenzo) [inline src/ORM/Behavior/TreeBehavior.php]: `IdentiferExpression` does not care about configuration, it is coded as a no-op when no quoting is enabled

Comment (lorenzo) [inline src/ORM/Behavior/TreeBehavior.php]: Would it make sense to override `config()` for this?

Comment (ADmad) [inline src/ORM/Behavior/TreeBehavior.php]: If there's no better option then yes. One should definitively be able to change the fields config post instance creation.

Comment (lorenzo) [inline src/ORM/Behavior/TreeBehavior.php]: I fail to see the practical approach to doing that, but I'll see what I can do

Comment (skie) [inline src/ORM/Behavior/TreeBehavior.php]: It is not really trivial to override method inside trait. May be more sense implement getter method that retruns wrapped left or right based of its argument on behavior level?

Comment (lorenzo) [inline src/ORM/Behavior/TreeBehavior.php]: As @skie said, it is almost impossible to override config() for this purpose. It is either this, or creating the expression object on each method

Comment (ADmad) [inline src/ORM/Behavior/TreeBehavior.php]: Then we should add new getter / setter for setting the fields at runtime.

**Comment by lorenzo** [comment]
I haven't had the time to finish this. I will try to add the getters soon

**Comment by markstory** [comment]
Sorry @lorenzo I thought this was done :cry:

**Pull request merged**